pkgs/dart_model/lib/src/scopes.dart

```diff
@@ -118,6 +123,47 @@ enum Scope {
 class _ScopeData {
   final Scope type;
   final JsonBufferBuilder? buffer;
+  final MacroScope? macro;
+
+  _ScopeData._(this.type, this.buffer, [this.macro]);
+}
+
+/// Data attached to [Scope.macro] environments to make resolved queries
```

Comment:
Nitpick: environments->scopes so we always refer to scopes as ... scopes :)

I often catch myself using synonyms in comments because it's better writing style, but then I realize what we really want is to have a small and precise set of nouns in play ;)